backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,360 @@
+	}
+
+	private boolean handleWorkPackages(@NonNull final Properties ctx, @NonNull final List<I_C_Queue_WorkPackage> workPackages)
+	{
+		if (workPackages.isEmpty())
+		{
+			return false;
+		}
+
+		boolean handledAllWorkPackages = true;
+
+		for (final I_C_Queue_WorkPackage workPackage : workPackages)
+		{
+			if (!isValid(workPackage))
+			{
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final Optional<IQueueProcessor> queueProcessorOptional = getAvailableQueueProcessorForWorkPackage(workPackage);
+
+			if (!queueProcessorOptional.isPresent())
+			{
+				logger.warn("*** processWorkPackages: No available QueueProcessor found for workPackageId: {}, c_queue_workpackage.c_queue_packageprocessor_id: {}",
+							workPackage.getC_Queue_WorkPackage_ID(), workPackage.getC_Queue_PackageProcessor_ID());
+
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final IQueueProcessor queueProcessor = queueProcessorOptional.get();
+
+			final Optional<I_C_Queue_WorkPackage> workPackageWithOwnCtx = getWorkPackageWithOwnContext(queueProcessor, workPackage, ctx);
+
+			if (!workPackageWithOwnCtx.isPresent())
+			{
+				logger.warn("*** processWorkPackages: Couldn't assign a new Ctx to C_Queue_WorkPackage: {}", workPackage.getC_Queue_WorkPackage_ID());
+
+				WorkPackageLockHelper.unlockNoFail(workPackage);
+				handledAllWorkPackages = false;
+				continue;
+			}
+
+			final boolean successfullyHandled = handleWorkPackageProcessing(queueProcessor, workPackageWithOwnCtx.get());
+
+			handledAllWorkPackages = handledAllWorkPackages && successfullyHandled;
+		}
+
+		return handledAllWorkPackages;
+	}
+
+	@NonNull
+	private List<I_C_Queue_WorkPackage> pollAndLockWorkPackages(@NonNull final Properties ctx, @NonNull final List<IQueueProcessor> queueProcessors)
+	{
+		final IQuery<I_C_Queue_WorkPackage> queueProcessorWPQueriesAggregator = queryBL.createQueryBuilder(I_C_Queue_WorkPackage.class)
+				//dev-note: workaround to be able to union multiple queries without applying 'limit' and 'order by'  to the end result
+				.addEqualsFilter(I_C_Queue_WorkPackage.COLUMNNAME_C_Queue_WorkPackage_ID, -1)
+				.create();
+
+		final int numberOfWorkPackagesPerProcessor = 1;
+
+		for (final IQueueProcessor queueProcessor : queueProcessors)
+		{
+			final IQuery<I_C_Queue_WorkPackage> workPackageQuery = queueProcessor.getQueue().createQuery(ctx, numberOfWorkPackagesPerProcessor);
+
+			final IQuery<I_C_Queue_WorkPackage> nonLockedWorkPackageQuery = lockManager.addNotLockedClause(workPackageQuery);
+
+			queueProcessorWPQueriesAggregator.addUnion(nonLockedWorkPackageQuery, true);
+		}
+
+		return lockManager.retrieveAndLockMultipleRecords(queueProcessorWPQueriesAggregator, I_C_Queue_WorkPackage.class);
+	}
+
+	@NonNull
+	private Optional<IQueueProcessor> getAvailableQueueProcessorForWorkPackage(@NonNull final I_C_Queue_WorkPackage workPackage)
```

Comment:
I would use QueuePackageProcessorId as a parameter instead of I_C_Queue_WorkPackage workPackage,
just to have a clear separation of who does what,
and to pass as less info as possible.